components/apimgt/org.wso2.carbon.apimgt.rest.api/src/main/java/org/wso2/carbon/apimgt/rest/api/interceptors/ValidationInInterceptor.java

```diff
@@ -0,0 +1,131 @@
+import javax.validation.ValidatorFactory;
+import javax.validation.executable.ExecutableValidator;
+
+public class ValidationInInterceptor extends AbstractPhaseInterceptor<Message> {
+    private Log log = LogFactory.getLog(getClass());
+    private Validator validator = null; //validator interface is thread-safe
+    
+    public ValidationInInterceptor() {
+        super(Phase.PRE_INVOKE);
+        ValidatorFactory defaultFactory = Validation.buildDefaultValidatorFactory();
+        validator = defaultFactory.getValidator();
+        if (validator == null){
+            log.warn("Bean Validation provider could not be found, no validation will be performed");
+        } else {
+            log.info("Validation In-Interceptor initialized successfully");
+        }
+    }
+
+    public void handleMessage(Message message) {
+
+        final OperationResourceInfo operationResource = message.getExchange().get(OperationResourceInfo.class);
+        if (operationResource == null) {
+            log.info("OperationResourceInfo is not available, skipping validation");
+            return;
+        }
+
+        final ClassResourceInfo classResource = operationResource.getClassResourceInfo();
+        if (classResource == null) {
+            log.info("ClassResourceInfo is not available, skipping validation");
+            return;
+        }
+
+        final ResourceProvider resourceProvider = classResource.getResourceProvider();
+        if (resourceProvider == null) {
+            log.info("ResourceProvider is not available, skipping validation");
+            return;
+        }
+
+        final List<Object> arguments = MessageContentsList.getContentsList(message);
+        final Method method = operationResource.getAnnotatedMethod();
+        final Object instance = resourceProvider.getInstance(message);
+        if (method != null && arguments != null) {
+            //validate the parameters(arguments) over the invoked method
+            validate(method, arguments.toArray(), instance);
+
+            //validate the fields of each argument
+            for (Object arg : arguments) {
+                if (arg != null)
+                    validate(arg);
+            }
+        }
+    }
+
+    public <T> void validate(final Method method, final Object[] arguments, final T instance) {
+        try {
+            if (validator == null) {
+                log.warn("Bean Validation provider could not be found, no validation will be performed");
+                return;
+            }
+
+            ExecutableValidator methodValidator = validator.forExecutables();
+            Set<ConstraintViolation<T>> violations = methodValidator.validateParameters(instance,
+                    method, arguments);
+
+            if (!violations.isEmpty()) {
+                throw new ConstraintViolationException(violations);
+            }
+
+        } catch (ValidationException ex) {
+            throw new InternalServerErrorException(ex);
```

Comment:
Fixed in 1528 (https://github.com/wso2/carbon-apimgt/commit/ab4f846a67ce38c739e8c44dbb462ba844e65bb6)